### DISCUSSION TITLE
New Error launching Mirth Connect Administrator, using Mirth Connect: 2.2.1.5861

### DISCUSSION THREAD

**Discussion by mogoodrich** [nextgenhealthcare/connect#4792] [Q&A]
Hello all!

We've been using Mirth Connect (version 2.2.1.15861) for years without issues, but I'm recently running into an error when trying to connect to a Mirth Connect server using the Mirth Connect Administrator (which I launched via the Mirth Connect Administrator Launcher) running on a Ubuntu 16.04 machine.

The log-in hangs on the "Please wait: Loading extensions..."  (screenshot attached below) and I'm seeing the following in the error Java Console.  Any thoughts?  I download and extracted a completely new version of the Mirth Connect Administrator Launcher in case that was part of the problem, but am still seeing the same issue.

If anyone has any ideas, it would be much appreciated!

Thanks,
Mark

```
Using JRE version 1.8.0_181-b13 Java HotSpot(TM) 64-Bit Server VM
Java home directory = /home/mgoodrich/Desktop/PACS/mirthadministrator/Mirth Connect Administrator Launcher/jre
User home directory = /home/mgoodrich
----------------------------------------------------
c:  Clear console window
s:  Dump system properties
----------------------------------------------------
java.lang.NullPointerException
	at com.mirth.connect.client.ui.Frame.alertException(Frame.java:1056)
	at com.mirth.connect.client.ui.Frame.alertException(Frame.java:1027)
	at com.mirth.connect.client.ui.Frame.refreshCodeTemplates(Frame.java:3942)
	at com.mirth.connect.client.ui.Frame.setupFrame(Frame.java:349)
	at com.mirth.connect.client.ui.Mirth.<init>(Mirth.java:58)
	at com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:419)
	at com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:399)
	at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at javax.swing.SwingWorker.run(SwingWorker.java:334)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

![2021-10-14_15-00](https://user-images.githubusercontent.com/1163015/137380988-e4dc2019-d702-441b-a4d5-ea9728be5d38.png)

Update:

I was able to run on a Windows machine and got what could be a more useful error message:

```
Unable to load extensions
com.mirth.connect.client.core.ServerConnection.executePostMethod(ServerConnection.java:108)
com.mirth.connect.client.core.Client.getConnectorMetaData(Client.java:262)
com.mirth.connect.client.ui.Frame.loadExtensionMetaData(Frame.java:433)
com.mirth.connect.client.ui.Frame.setupFrame(Frame.java:346)
com.mirth.connect.client.ui.Mirth.<init>(Mirth.java:58)
com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:419)
com.mirth.connect.client.ui.LoginPanel$8.doInBackground(LoginPanel.java:399)
javax.swing.SwingWorker$1.call(Unknown Source)
java.util.concurrent.FutureTask.run(Unknown Source)
javax.swing.SwingWorker.run(Unknown Source)
java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
java.lang.Thread.run(Unknown Source)
```

**Comment by pacmano1** [comment]
I can't get this NOT to work on assorted mixes of the server and desktop OS.  Do you have custom extensions?   I am not running java 8 on the desktop though, rather openjdk 11 or 13.   Server side I tested with was openjdk 8.

As you know you are on an old version.  Released in 2012.    So I only tested on VMs.   About time to upgrade, no?

I assume you cleared the java cache which is local the the admin launcher, not the "standard" one on your OS..

**Comment by jonbartels** [comment]
Anything on the server end in `mirth.log`?

Are you launching MC using plain `javaws` or using the MC Launcher?

I don't have 2.2.1 checked out at the moment but the stack trace suggests that its having an issue refreshing code templates *or* its trying to throw an alert and can't. What is the code around `Frame.java` line 1056 for 2.2.1?

`	at com.mirth.connect.client.ui.Frame.alertException(Frame.java:1056)
	at com.mirth.connect.client.ui.Frame.alertException(Frame.java:1027)
	at com.mirth.connect.client.ui.Frame.refreshCodeTemplates(Frame.java:3942)`

**Comment by mogoodrich** [comment]
Thanks @jonbartels  @pacmano1  for the quick response!

To answer your questions:

* No custom extensions that I know of
* I'm using the Mirth Connection Administrator Launcher
* I wasn't sure how to clear the cache local to the launcher... but I downloaded two different instances of the launcher on my Linux machine, and then downloaded on a Windows machine and had the same issue, so I was assuming it wasn't a cache issue?
* Nothing in the mirth.log on the server... I pumped the logging level up the DEBUG
* I think the "alertException" is an exception with rendering the alert message, I think the underlying place it is failing is from the logs I retrieved when running on the Windows machine:

https://github.com/nextgenhealthcare/connect/blob/2.2.1/server/src/com/mirth/connect/client/core/ServerConnection.java#L108

Which is coming from here:

https://github.com/nextgenhealthcare/connect/blob/2.2.1/server/src/com/mirth/connect/client/core/Client.java#L262

And to the upgrade question... yeah, it's been on my radar for a while now that we should upgrade, but this is a legacy application that has been chugging along nicely for years, so it keeps getting de-prioritized, but that will likely be the next step if there's no obvious answer to what is going on.

Is there any way to turn on any more verbose logging on the client side?  I'd love to see what the underlying exception triggering the Client Exception is.

I'm also trying to think about what might have changed recently on the server, but nothing coming to mind.

Thanks again!
Mark

**Reply by jonbartels** [reply]
I think your reference to https://github.com/nextgenhealthcare/connect/blob/2.2.1/server/src/com/mirth/connect/client/core/Client.java#L262 is key to figuring out the issue. Since that is a Client call, you can call the MC API for each extension or code template and see what it returns. If it fails on a "get all" call or on getting a specific object, that starts to pinpoint the issue.

As to caching the Launcher is designed so you don't have to clear the cache. That is it's major feature.

**Reply by mogoodrich** [reply]
Thanks @jonbartels , that makes sense.   I've really just been a user of Mirth, I haven't accessed the API... I can try and reverse engineer via the client code, but is there any documentation on how to make a request of the MC API?  Thanks!

**Reply by jonbartels** [reply]
I'm on mobile please excuse my brevity. 

Yes there is API doco but for 2.2 your best bet is the forums, https://forums.mirthproject.io/

I think in 2.2 you have to get a JWT cookie first then make API calls. You can see in the cited code it's just HTTP and XML.
